Tribler/community/tunnel/processes/processmanager.py

```diff
@@ -0,0 +1,186 @@
+
+    def get_worker_count(self):
+        """Return the active amount of workers
+
+        :return: the current worker count
+        :rtype: int
+        """
+        return len(self.pool.keys())
+
+    def _create_worker(self, callback):
+        """Create a single worker, without adding it to the pool
+
+        :param callback: the callback for when the worker is created
+        :type callback: func
+        :returns: None
+        """
+        key_pair =\
+            self.session.get_multichain_permid_keypair_filename()
+        is_exit_node =\
+            self.session.get_tunnel_community_exitnode_enabled()
+        test_mode = self.session.get_tunnel_community_test_pooled()
+
+        process = TunnelProcess(
+            self.community if self.community
+            else self.session.lm.tunnel_community)
+
+        def on_created(proc):
+            callback(proc)
+            proc.create(key_pair, is_exit_node, test_mode)
+        process.started.addCallback(on_created)
+
+    def _remove_workers(self, amount):
+        """Remove some amount of workers
+
+        :param amount: the amount of workers to remove
+        :type amount: int
+        :returns: None
+        """
+        to_remove = sample(self.pool.keys(), amount)
+        for worker in to_remove:
+            self.pool[worker].end()
+            self.pool.pop(worker)
+        self.circuit_map = {
+            k:v for k, v in self.circuit_map.iteritems()
+            if v in to_remove}
+
+    def monitor_infohashes(self, infohashes):
+        """Call monitor_infohashes on all workers in the pool
+
+        :param infohashes: the infohashes the workers need to monitor
+        :type infohashes: [(str, int, int)]
+        :returns: None
+        """
+        for worker in self.pool.values():
+            worker.monitor_infohashes(infohashes)
+
+    def send_data(self, candidates, circuit_id, dest_address,
+                  source_address, data):
+        """Call send_data() on the worker assigned to circuit_id
+
+        :param candidates: the candidates to use
+        :type candidates: Tribler.dispersy.candidate.Candidate
+        :param circuit_id: the circuit id to send over
+        :type circuit_id: long
+        :param dest_address: the destination address to send to
+        :type dest_address: (str, int)
+        :param source_address: the source address to send from
+        :type source_address: (str, int)
+        :param data: the data to send
+        :type data: str
+        :returns: None
+        """
+        if circuit_id in self.circuit_map:
+            worker = self.pool[self.circuit_map[circuit_id]]
+            worker.send_data([cd.sock_addr for cd in candidates],
+                             circuit_id,
+                             dest_address,
+                             source_address,
+                             data)
+        else:
+            logging.warning("Could not find worker registered for " +
```

Comment:
It can happen that `create_circuit()` returns a `circuit_id` which is not yet known for any circuit shared by the _PooledTunnelCommunity.circuits_ `SyncDict`, even though it exists. This is a brute-force 'fix' for a situation that will resolve itself momentarily. LibTorrent does not respond very well to any fancy fixes for this (like delaying or dropping the data).